components/cluster/command/deploy.go

```diff
@@ -270,37 +276,80 @@ func deploy(clusterName, clusterVersion, topoFile string, opt deployOptions) err
 		// log dir will always be with values, but might not used by the component
 		logDir := clusterutil.Abs(globalOptions.User, inst.LogDir())
 		// Deploy component
+		// prepare deployment server
 		t := task.NewBuilder().
 			UserSSH(inst.GetHost(), inst.GetSSHPort(), globalOptions.User, gOpt.SSHTimeout).
 			Mkdir(globalOptions.User, inst.GetHost(),
 				deployDir, logDir,
 				filepath.Join(deployDir, "bin"),
 				filepath.Join(deployDir, "conf"),
 				filepath.Join(deployDir, "scripts")).
-			Mkdir(globalOptions.User, inst.GetHost(), dataDirs...).
-			CopyComponent(
-				inst.ComponentName(),
+			Mkdir(globalOptions.User, inst.GetHost(), dataDirs...)
+
+		// copy dependency component if needed
+		switch inst.ComponentName() {
+		case spec.ComponentTiSpark:
+			sparkSubPath := spec.ComponentSubDir(spec.ComponentSpark,
+				spec.ComponentVersion(spec.ComponentSpark, version))
+			t = t.CopyComponent(
+				spec.ComponentSpark,
 				inst.OS(),
 				inst.Arch(),
-				version,
+				spec.ComponentVersion(spec.ComponentSpark, version),
 				inst.GetHost(),
 				deployDir,
-			).
-			InitConfig(
-				clusterName,
-				clusterVersion,
-				inst,
-				globalOptions.User,
-				opt.ignoreConfigCheck,
-				meta.DirPaths{
-					Deploy: deployDir,
-					Data:   dataDirs,
-					Log:    logDir,
-					Cache:  spec.ClusterPath(clusterName, spec.TempConfigPath),
-				},
-			).
-			BuildAsStep(fmt.Sprintf("  - Copy %s -> %s", inst.ComponentName(), inst.GetHost()))
-		deployCompTasks = append(deployCompTasks, t)
+			).Shell( // spark is under a subdir, move it to deploy dir
+				inst.GetHost(),
+				fmt.Sprintf(
+					"cp -rf %[1]s %[2]s/ && cp -rf %[3]s/* %[2]s/ && rm -rf %[1]s %[3]s",
```

Comment:
I think it's no matter that we retain the `filepath.Join(deployDir, "bin", sparkSubPath)` and `filepath.Join(deployDir, sparkSubPath)` directories.

Reply:
They're mixed with decompressed files and may cause confuses to users.